designs/move-cluster-specific-code-out-of-manager.md

```diff
@@ -0,0 +1,209 @@
+	// Runnable to the manager via Add method.
+	AddMetricsExtraHandler(path string, handler http.Handler) error
+
+	// AddHealthzCheck allows you to add Healthz checker
+	AddHealthzCheck(name string, check healthz.Checker) error
+
+	// AddReadyzCheck allows you to add Readyz checker
+	AddReadyzCheck(name string, check healthz.Checker) error
+
+	// Start starts all registered Controllers and blocks until the Stop channel is closed.
+	// Returns an error if there is an error starting any controller.
+	// If LeaderElection is used, the binary must be exited immediately after this returns,
+	// otherwise components that need leader election might continue to run after the leader
+	// lock was lost.
+	Start(<-chan struct{}) error
+
+	// GetWebhookServer returns a webhook.Server
+	GetWebhookServer() *webhook.Server
+}
+```
+
+Furthermore, during startup, the `Manager` will use type assertion to find `ClusterConnector`s
+to be able to start their caches before anything else:
+
+```go
+if cc, isClusterConnector:= runnable.(clusterconnector.ClusterConnector); isClusterConnector {
+	m.caches = append(m.caches, cc.GetCache())
+}
+```
+
+```go
+for idx := range cm.caches {
+	go func(idx int) {cm.caches[idx].Start(cm.internalStop)}
+}
+
+for _, cache := range cm.caches {
+	cache.WaitForCacheSync(cm.internalStop)
+}
+
+// Start all other runnables
+```
+
+## Example
+
+Below is a sample `reconciler` that will create a secret in a `mirrorCluster` for each
+secret found in `referenceCluster` if none of that name already exists. To keep the sample
+short, it won't compare the contents of the secrets.
+
+```go
+type secretMirrorReconciler struct {
+	referenceClusterClient, mirrorClusterClient client.Client
+}
+
+func (r *secretMirrorReconciler) Reconcile(r reconcile.Request)(reconcile.Result, error){
+	s := &corev1.Secret{}
+	if err := r.referenceClusterClient.Get(context.TODO(), r.NamespacedName, s); err != nil {
+		if kerrors.IsNotFound{ return reconcile.Result{}, nil }
+		return reconcile.Result, err
+	}
+
+	if err := r.mirrorClusterClient.Get(context.TODO(), r.NamespacedName, &corev1.Secret); err != nil {
+		if !kerrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+
+		mirrorSecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{Namespace: s.Namespace, Name: s.Name},
+			Data: s.Data,
+		}
+		return reconcile.Result{}, r.mirrorClusterClient.Create(context.TODO(), mirrorSecret)
+	}
+
+	return nil
+}
+
+func NewSecretMirrorReconciler(mgr manager.Manager, mirrorConnector clusterconnector.ClusterConnector) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// Watch Secrets in the reference cluster
+		For(&corev1.Secret{}).
+		// Watch pods in the mirror cluster
+		Watches(
+			source.NewKindWithCache(&corev1.Secret{}, mirrorCluster.GetCache()),
+			&handler.EnqueueRequestForObject{},
+		).
+		Complete(&secretMirrorReconciler{
+			referenceClusterClient: mgr.GetClient(),
+			mirrorClusterClient:    mirrorCluster.GetClient(),
+		})
+	}
+}
```

Comment:
@alvaroaleman I like this approach, currently, we can do something like the following to achieve the same feature with only one manager:
```go
func  (r *secretMirrorReconciler) SetupWithManager(mgr manager.Manager, mirrorClusterCache cache.Cache) error {
	return ctrl.NewControllerManagedBy(mgr).
		// Watch Secrets in the reference cluster
		For(&corev1.Secret{}).
		// Watch pods in the mirror cluster
		Watches(
			source.NewKindWithCache(&corev1.Secret{}, mirrorClusterCache),
			&handler.EnqueueRequestForObject{},
		).
		Complete(r)
	}
}
```
in `main.go` we can do:
```go
func main(){
	mgr, err := manager.New(cfg1, manager.Options{})
	if err != nil {
		panic(err)
	}
	mirrorMapper, err := apiutil.NewDiscoveryRESTMapper(mirrorCfg)
	if err != nil {
		panic(err)
	}
	mirrorClient, err := client.New(mirrorCfg, client.Options{
		Scheme: mirrorScheme,
		Mapper: mirrorMapper,
	})
	if err != nil {
		panic(err)
	}
	mirrorCache, err := cache.New(mirrorCfg, cache.Options{
		Scheme: mirrorScheme,
		Mapper: mirrorMapper,
	})
	if err = mgr.Add(mirrorCache); err != nil {
		panic(err)
	}
	if err := mgr.Add(mirrorConnector); err != nil {
		panic(err)
	}
	if err := (&SecretMirrorReconciler{
	        ReferenceClusterClient: mgr.GetClient(),
		MirrorClusterClient:    mirrorCluster.GetClient(),
	}).SetupWithManager(mgr, mirrorCache); err != nil {
		panic(err)
	}
	if err := mgr.Start(signals.SetupSignalHandler()); err != nil {
		panic(err)
	}
}
```
Which requires a lot of work & setup, if we have the connector implemented, should be much easier to do cross-cluster stuff :)

Reply:
@jiachengxu Yeah I am aware that this is possible, but it is very inconvenient to use and until recently it required rebuilding the `DefaultNewClient` because it wasn't exported (your example uses a api-backed `Reader` for the mirror cluster even though you have a cache)